index.js

```diff
@@ -522,11 +555,48 @@ class KeyringController extends EventEmitter {
         return { type, data };
       }),
     );
-    const encryptedString = await this.encryptor.encrypt(
-      this.password,
-      serializedKeyrings,
-    );
-    this.store.updateState({ vault: encryptedString });
+
+    let vault;
+    let newExportedKeyString;
+
+    if (this.cacheEncryptionKey) {
+      if (this.password) {
+        const { vault: newVault, exportedKeyString } =
+          await this.encryptor.encryptWithDetail(
+            this.password,
+            serializedKeyrings,
+          );
+
+        vault = newVault;
+        newExportedKeyString = exportedKeyString;
+      } else if (encryptionKey) {
+        const key = await this.encryptor.importKey(encryptionKey);
+        const vaultJSON = await this.encryptor.encryptWithKey(
+          key,
+          serializedKeyrings,
+        );
+        vaultJSON.salt = this.encryptionSalt;
+        vault = JSON.stringify(vaultJSON);
+      }
+    } else {
+      vault = await this.encryptor.encrypt(this.password, serializedKeyrings);
+    }
+
+    if (!vault) {
+      throw new Error('Cannot persist vault without vault information');
+    }
+
+    this.store.updateState({ vault });
+
+    // The keyring updates need to be announced before updating the encryptionKey
+    // so that the updated keyring gets propagated to the extension first.
+    // Not calling _updateMemStoreKeyrings results in the wrong account being selected
+    // in the extension.
+    await this._updateMemStoreKeyrings();
+    if (newExportedKeyString) {
+      this.memStore.updateState({ encryptionKey: newExportedKeyString });
```

Comment:
Hmm. Should we keep the `encryptionSalt` in memory alongside the `encryptionKey`? That would help us verify whether the `encryptionKey` matches what was used to encrypt the vault.

Every time a password is used to encrypt the vault, a new encryption key + salt is generated. Once that new encrypted vault is persisted, the previous `encryptionKey` becomes obsolete - it was generated using a different salt, so it will not work to decrypt the vault anymore.

The only scenario where this might happen is if the vault is persisted 3 lines up from here, but then the service worker dies before setting the encryption key in session storage. In that scenario, IF there was already an encryption key set previously, we'd have an invalid key in session storage and the next login attempt would fail.

However if we set the salt in session storage as well, we can compare the session salt to the vault salt to validate the key prior to using it. Then we can bump the user to the login screen in this edge case, and avoid leaving them stuck at the login screen.